fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java

```diff
@@ -1501,68 +1497,71 @@ private void processMvRelatedMeta() {
-                                + "please update FeConstants.starrocks_meta_version bigger or equal to {} and restart.",
-                        starrocksMetaVersion);
-                System.exit(-1);
-            }
-            newChecksum ^= starrocksMetaVersion;
-            MetaContext.get().setStarRocksMetaVersion(starrocksMetaVersion);
+            checksum ^= dis.readInt();
+            starrocksMetaVersion = dis.readInt();
+            checksum ^= starrocksMetaVersion;
         } else {
-            // when flag is positive, this is community image structure
-            int metaVersion = flag;
-            if (metaVersion > FeConstants.META_VERSION) {
-                LOG.error("invalid meta data version found, cat not bigger than FeConstants.meta_version."
-                                + "please update FeConstants.meta_version bigger or equal to {} and restart.",
-                        metaVersion);
-                System.exit(-1);
-            }
-            MetaContext.get().setMetaVersion(metaVersion);
+            // when flag is positive, this is new version format
+            starrocksMetaVersion = flag;
         }
 
-        long replayedJournalId = dis.readLong();
-        newChecksum ^= replayedJournalId;
+        if (starrocksMetaVersion > FeConstants.STARROCKS_META_VERSION) {
+            LOG.error("invalid meta data version found, cat not bigger than FeConstants.starrocks_meta_version."
+                            + "please update FeConstants.starrocks_meta_version bigger or equal to {} and restart.",
+                    starrocksMetaVersion);
+            System.exit(-1);
+        }
+        MetaContext.get().setStarRocksMetaVersion(starrocksMetaVersion);
 
-        long catalogId = dis.readLong();
-        newChecksum ^= catalogId;
-        idGenerator.setId(catalogId);
+        return checksum;
+    }
 
-        if (MetaContext.get().getMetaVersion() >= FeMetaVersion.VERSION_32) {
-            isDefaultClusterCreated = dis.readBoolean();
+    public long loadHeader(DataInputStream dis, long checksum) throws IOException {
+        if (GlobalStateMgr.getCurrentStateStarRocksMetaVersion() >= StarRocksFEMetaVersion.VERSION_4) {
+            return loadHeaderV2(dis, checksum);
+        } else {
+            return loadHeaderV1(dis, checksum);
         }
+    }
+
+    public long loadHeaderV1(DataInputStream dis, long checksum) throws IOException {
+        long replayedJournalId = dis.readLong();
+        checksum ^= replayedJournalId;
+
+        long batchEndId = dis.readLong();
+        checksum ^= batchEndId;
+        idGenerator.setId(batchEndId);
+
+        isDefaultClusterCreated = dis.readBoolean();
 
         LOG.info("finished replay header from image");
-        return newChecksum;
+        return checksum;
+    }
+
+    public long loadHeaderV2(DataInputStream dis, long checksum) throws IOException {
+        ImageHeader header = GsonUtils.GSON.fromJson(Text.readString(dis), ImageHeader.class);
```

Comment:
we don't need `replayedJournalId` here any more?

Reply:
yes, it's useless now.